goalie/go_mesh_seq.py

```
@@ -43,15 +43,18 @@ def read_forms(self, forms_dictionary):
         :type forms_dictionary: :class:`dict`
         """
         for fieldname, form in forms_dictionary.items():
-            if fieldname not in self.field_functions:
-                raise ValueError(
-                    f"Unexpected field '{fieldname}' in forms dictionary."
-                    f" Expected one of {list(self.field_metadata.keys())}."
-                )
-            if not isinstance(form, ufl.Form):
-                raise TypeError(
-                    f"Expected a UFL form for field '{fieldname}', not '{type(form)}'."
-                )
+            field = self.field_metadata[fieldname]
+            if field.solved_for:
+                if fieldname not in self.field_functions:
+                    raise ValueError(
+                        f"Unexpected field '{fieldname}' in forms dictionary."
+                        f" Expected one of {list(self.field_metadata.keys())}."
+                    )
+                if not isinstance(form, ufl.Form):
+                    raise TypeError(
+                        f"Expected a UFL form for field '{fieldname}', not"
+                        f" '{type(form)}'."
+                    )
```

```suggestion
        expected_fields = [
            fieldname
            for fieldname, field in self.field_metadata.items()
            if field.solved_for
        ]
        for fieldname, form in forms_dictionary.items():
            if fieldname not in expected_fields:
                raise ValueError(
                    f"Unexpected field '{fieldname}' in forms dictionary."
                    f" Expected one of {expected_fields}."
                )
            field = self._get_field_metadata(fieldname)
            if not field.solved_for:
                raise ValueError(
                    f"Unexpected field '{fieldname}' in forms dictionary."
                    f" Field '{fieldname}' is not solved for."
                )
            if not isinstance(form, ufl.Form):
                raise TypeError(
                    f"Expected a UFL form for field '{fieldname}', not '{type(form)}'."
                )
```

I think it should look like this. If we had `fields = [Field("c"), Field("u", solved_for=False)]` and then tried to do `mesh_seq.read_forms({"c": F, "u": F, "v": F})` we would want errors raised for `u` and `v`. It looks like `v` would be silently skipped in your change.

---

Addressed in cf1d572b41408ee18ffd014b5d2ec0b372074e6f.

---

Sorry, but this would still silently skip `u` in my example above. And since we have
```
                raise ValueError(
                    f"Unexpected field '{fieldname}' in forms dictionary."
                    f" Expected one of {list(self.field_metadata.keys())}."
                )
```
it would still say `Expected one of ['c', 'u']`.

---

Feel free to push the required fix if you like - I won't have time to look at this til at least tomorrow.

---

Perfect, thanks @ddundo!